File: src/atlas_map/src/scene/auvTiles.js
import { KX, KZ, toX, toZ } from "./geo.js";

export function decode(int16, S, zScale, zOffset) {
  const m = new Float32Array(S * S);
  for (let r = 0; r < S; r++) { let acc = 0; for (let c = 0; c < S; c++) { acc += int16[r * S + c]; m[r * S + c] = acc / zScale + zOffset; } }
  return m;
}

export function wanted(levels, target, camDist, tileCenter) {
  const w = [new Set(levels[0].have), new Set(), new Set()];
  const near = (k, key, R) => { const [cx, cz] = tileCenter(k, key); return Math.hypot(cx - target.x, cz - target.z) < R; };
  if (camDist < 4) {
    const R2 = Math.min(1.2, Math.max(0.35, camDist * 0.35));
    for (const key of levels[2].have) if (near(2, key, R2)) {
      w[2].add(key);
      const [ty, tx] = key.split("_").map(Number);
      w[1].add(`${Math.floor(ty / 4)}_${Math.floor(tx / 4)}`);
    }
  }
  if (camDist < 25) {
    const R1 = Math.min(4, Math.max(1.5, camDist * 0.5));
    for (const key of levels[1].have) if (near(1, key, R1)) w[1].add(key);
  }
  return w;
}

export function tileArrays(h, S, lon0, lat0, cellDeg) {
  const dx = cellDeg * KX, dz = cellDeg * KZ, n = S * S, ring = 4 * (S - 1);
  const H = (r, c) => h[Math.min(S - 1, Math.max(0, r)) * S + Math.min(S - 1, Math.max(0, c))];
  const positions = new Float32Array((n + ring) * 3), elev = new Float32Array(n + ring),
        grad = new Float32Array((n + ring) * 2), drop = new Float32Array(n + ring);
  for (let r = 0, i = 0; r < S; r++) for (let c = 0; c < S; c++, i++) {
    positions[i * 3] = toX(lon0 + (c + 0.5) * cellDeg); positions[i * 3 + 2] = toZ(lat0 - (r + 0.5) * cellDeg);
    elev[i] = h[i];
    grad[i * 2] = (H(r, c + 1) - H(r, c - 1)) / (2 * dx); grad[i * 2 + 1] = (H(r + 1, c) - H(r - 1, c)) / (2 * dz);
  }
  const index = [];
  for (let r = 0; r < S - 1; r++) for (let c = 0; c < S - 1; c++) {
    const a = r * S + c, b = a + 1, d = a + S, e = d + 1; index.push(a, d, b, b, d, e);
  }
  const edge = [];
  for (let c = 0; c < S - 1; c++) edge.push(c);
  for (let r = 0; r < S - 1; r++) edge.push(r * S + S - 1);
  for (let c = S - 1; c > 0; c--) edge.push((S - 1) * S + c);
  for (let r = S - 1; r > 0; r--) edge.push(r * S);
  edge.forEach((v, k) => {
    const j = n + k;
    positions[j * 3] = positions[v * 3]; positions[j * 3 + 2] = positions[v * 3 + 2];
    elev[j] = elev[v]; grad[j * 2] = grad[v * 2]; grad[j * 2 + 1] = grad[v * 2 + 1]; drop[j] = 25;
  });
  for (let k = 0; k < edge.length; k++) {
    const a = edge[k], b = edge[(k + 1) % edge.length], a2 = n + k, b2 = n + ((k + 1) % edge.length);
    index.push(a, a2, b, b, a2, b2);
  }
  return { positions, elev, grad, drop, index: new Uint32Array(index) };
}

// [lon0, lat0, lon1, lat1] (west, north, east, south) that a level's tiles cover when drawn. Vertices sit at
// cell centers, so this is the tile grid shifted half a cell east and south of the survey's corner.
export function coverage(west, north, cellDeg, tileDeg, tilesX, tilesY) {
  const lon0 = west + cellDeg / 2, lat0 = north - cellDeg / 2;
  return [lon0, lat0, lon0 + tilesX * tileDeg, lat0 - tilesY * tileDeg];
}

File: src/atlas_map_data/auv_tiles.py
"""MBARI AUV bathymetry → level-of-detail tiles for the atlas (needs h5py + numpy; separate from the stdlib build).

  python -m atlas_map_data.auv_tiles --source <MBARI ... _AUVOverShip_Topo1mSq.grd> --out src/atlas_map/public/atlas/auv
"""
from __future__ import annotations

import argparse
import gzip
import json
import math
from pathlib import Path

import h5py
import numpy as np

CREDIT = "MBARI Axial Seamount bathymetry, cruise V2506: 1 m AUV survey merged over ship-based multibeam"
Z_OFFSET, Z_SCALE = -1000.0, 10.0
KX = 111.32 * math.cos(math.radians(45.94))
KZ = 111.13


def _encode(block: np.ndarray) -> bytes:
    q = np.clip(np.round((block - Z_OFFSET) * Z_SCALE), -32000, 32000).astype(np.int32)
    d = q.copy()
    d[:, 1:] = q[:, 1:] - q[:, :-1]
    if np.abs(d[:, 1:]).max(initial=0) > 32000:
        raise ValueError("delta overflow; raise Z_SCALE resolution or reduce stride")
    return gzip.compress(d.astype("<i2").tobytes(), 9)


def decode(data: bytes, cells: int, z_offset: float, z_scale: float) -> list[list[float]]:
    s = cells + 1
    d = np.frombuffer(gzip.decompress(data), dtype="<i2").reshape(s, s).astype(np.int64)
    return (np.cumsum(d, axis=1) / z_scale + z_offset).tolist()


def build(source: Path, out: Path, sites: list[tuple[float, float]], cells: int = 256,
          levels=((0, 16), (1, 4), (2, 1)), l2_radius_km: float = 1.2) -> dict:
    with h5py.File(source, "r") as f:
        lon, lat, z = f["lon"][:], f["lat"][:], f["z"]
        ny, nx = z.shape
        cell = float(lon[1] - lon[0])
        west, south = float(lon[0]) - cell / 2, float(lat[0]) - cell / 2
        north = south + ny * cell
        finest = max(level for level, _ in levels)

        def near_site(lon0, lat0, lon1, lat1):
            for plat, plon in sites:
                cx, cy = min(max(plon, lon0), lon1), min(max(plat, lat0), lat1)
                if math.hypot((cx - plon) * KX, (cy - plat) * KZ) <= l2_radius_km:
                    return True
            return False

        index = {"source": source.name, "credit": CREDIT, "west": west, "north": north, "cellDeg": cell,
                 "tileCells": cells, "zOffset": Z_OFFSET, "zScale": Z_SCALE, "nx": nx, "ny": ny, "levels": []}
        for level, stride in levels:
            lnx, lny = math.ceil(nx / stride), math.ceil(ny / stride)
            tnx, tny = math.ceil(lnx / cells), math.ceil(lny / cells)
            tile_deg = cells * stride * cell
            outdir = out / f"L{level}"
            outdir.mkdir(parents=True, exist_ok=True)
            tiles = []
            for ty in range(tny):
                lat1, lat0 = north - ty * tile_deg, north - (ty + 1) * tile_deg
                wanted = [tx for tx in range(tnx) if level != finest or
                          near_site(west + tx * tile_deg, lat0, west + (tx + 1) * tile_deg, lat1)]
                if not wanted:
                    continue
                r0 = ty * cells
                last = min(r0 + cells, lny - 1)
                lo, hi = max(ny - (last + 1) * stride, 0), ny - r0 * stride     # ascending-lat source rows
                strip = z[lo:hi, :].astype(np.float32)[::-1]                      # north-first
                if stride > 1:
                    h, w = (strip.shape[0] // stride) * stride, (nx // stride) * stride
                    rows_needed = last - r0 + 1
                    if h // stride < rows_needed:                                  # ragged south edge
                        strip = np.concatenate([strip, np.repeat(strip[-1:], rows_needed * stride - strip.shape[0], axis=0)])
                        h = rows_needed * stride
                    pooled = strip[:h, :w].reshape(h // stride, stride, w // stride, stride).mean(axis=(1, 3))
                    if w < nx:
                        pooled = np.concatenate([pooled, pooled[:, -1:]], axis=1)
                else:
                    pooled = strip
                for tx in wanted:
                    block = pooled[:cells + 1, tx * cells: tx * cells + cells + 1]
                    if block.size == 0 or np.isnan(block).all():
                        continue
                    pr, pc = cells + 1 - block.shape[0], cells + 1 - block.shape[1]
                    if pr or pc:
                        block = np.pad(block, ((0, pr), (0, pc)), mode="edge")
                    if np.isnan(block).any():
                        block = np.where(np.isnan(block), np.nanmean(block), block)
                    (outdir / f"{ty}_{tx}.bin.gz").write_bytes(_encode(block))
                    tiles.append([ty, tx])
            index["levels"].append({"level": level, "stride": stride, "cellDeg": cell * stride,
                                    "tilesX": tnx, "tilesY": tny, "tiles": tiles})
    out.mkdir(parents=True, exist_ok=True)
    (out / "index.json").write_text(json.dumps(index))
    return index


def main(argv: list[str] | None = None) -> int:
    from . import paths
    ap = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--source", type=Path, required=True)
    ap.add_argument("--out", type=Path, default=paths.BUNDLE / "auv")
    ap.add_argument("--sensors", type=Path, default=paths.BUNDLE / "sensors.json")
    args = ap.parse_args(argv)
    sensors = json.loads(args.sensors.read_text())["sensors"]
    sites = [(s["lat"], s["lon"]) for s in sensors if s.get("lat") is not None]
    ix = build(args.source, args.out, sites)
    for L in ix["levels"]:
        print(f"L{L['level']}: stride {L['stride']}, {len(L['tiles'])} tiles")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
